nemo_rl/models/generation/vllm.py

```
@@ -1378,43 +1416,90 @@ def generate_async(
+
+        # Start tasks for all worker generators
+        worker_tasks = [
+            asyncio.create_task(consume_worker_generator(i, proxy))
+            for i, proxy in enumerate(future_bundle)
+        ]
 
-        # Combine results from all tied worker groups
-        combined = BatchedDataDict.from_batches(
-            results, pad_value_dict={"output_ids": self.cfg["pad_token_id"]}
-        )
+        if not worker_tasks:
+            return
 
-        # Verify the output has all required fields
-        required_keys = [
-            "output_ids",
-            "generation_lengths",
-            "unpadded_sequence_lengths",
-            "logprobs",
-        ]
-        missing_keys = [key for key in required_keys if key not in combined]
-        if missing_keys:
-            raise ValueError(
-                f"Missing required keys for GenerationOutputSpec: {missing_keys}"
-            )
+        # Yield sample results as they become available from any worker
+        sample_count = 0
 
-        return combined
+        while finished_workers < total_workers:
+            try:
+                msg_type, item = await asyncio.wait_for(
+                    result_queue.get(), timeout=120.0
+                )
+            except asyncio.TimeoutError:
+                print(
+                    f"Timeout waiting for results. finished_workers: {finished_workers}/{total_workers}"
+                )
+                # Cancel all remaining tasks
+                for task in worker_tasks:
+                    if not task.done():
+                        task.cancel()
+                await asyncio.gather(*worker_tasks, return_exceptions=True)
+                raise RuntimeError("Timeout waiting for worker results")
+
+            if msg_type == "sample":
+                sample_count += 1
+                # Yield individual sample result immediately
+                yield item
+            elif msg_type == "error":
```

nit: can we also account for the `worker_done` message type in this if-block and then raise an exception if we see an unexpected one? i'm worried if someone changes the message type above it'll silently pass thru here